set up app tests

xianny opened the pull request:
**====== EDIT ======**
Revisited this and realised we needed more refactoring to allow dependency injection of testing mocks.

Pretty fiddly refactor, so I deployed to staging and confirmed smoke tests all pass. 

Changes:
- Cleaned up `/runners` and separated the logic into functions and executions. Removed the duplicate logic in `/index`
- Re-used `/runners` to set up the app in `/app.ts`
- `/index.ts` now uses `/app.ts` to start the API instead of duplicating the logic in `/runners`, so we can be sure that all entrypoints are running the same things
- tests use `/app.ts` to inject dependencies and start the API
- removed `/services/*_http_service.ts` because they aren't real services, they just set up routing. Moved the logic to `/routers`.


Questions:
- should we explicitly declare `/index` as the default entry point in package.json? we don't have one right now

**====== END EDIT ======**

Set up a basic test framework:
- set up CI and yarn scripts
- migrate 0x contracts
- extract express app to prepare for dependency injection
- test that GET /sra/orders returns 200 status

After merging this, will open more smaller PRs with more tests.

I tried splitting the tests into separate parallel CI jobs, but spinning up the env for each job actually takes more time (0:58 vs 4:03) When/if tests start taking longer, we can split up the jobs again. 
<img width="1211" alt="Screen Shot 2020-01-08 at 9 19 20 AM" src="https://user-images.githubusercontent.com/8582774/72000358-fb3d4080-31f7-11ea-8cf9-762c62501913.png">

dekz commented on test/app_test.ts:
```diff
@@ -0,0 +1,67 @@
+let web3Wrapper: Web3Wrapper;
+let provider: Web3ProviderEngine;
+let accounts: string[];
+// let contractAddresses: ContractAddresses;
```
```suggestion

```

dekz commented on test/app_test.ts:
```diff
@@ -0,0 +1,67 @@
+    before(async () => {
+        // start ganache and run contract migrations
+        const ganacheConfigs = {
+            total_accounts: 10,
```
I don't think you need this to be set

dekz commented on src/app.ts:
```diff
@@ -0,0 +1,70 @@
+        const orderStore = new OrderStoreDbAdapter(orderBookService);
+        const orderProvider = new OrderBookServiceOrderProvider(orderStore, orderBookService);
+        const orderBook = new Orderbook(orderProvider, orderStore);
+        const provider = providerUtils.createWeb3Provider(config.ETHEREUM_RPC_URL);
```
I'm not 100% certain if the `shouldUseInProcessGanache` also opens up ports for RPC access. We may need to inject the provider if we want to get the entire app to use the in process ganache.

None of the tests currently hit the RPC URL (as it's just get orders) but if they did I'm not certain how it would respond.

fragosti commented on src/app.ts:
```diff
@@ -0,0 +1,70 @@
+/**
+ * returns the app
+ */
+export async function getAppAsync(config: any = defaultConfig): Promise<express.Application> {
```
can we make the config typed but use `any` in the tests if it is even necessary? I don't see this being used anywhere.

xianny replied:
got rid of the `any`s

fragosti commented on src/db_connection.ts:
```diff
@@ -2,9 +2,10 @@ import { Connection, createConnection } from 'typeorm';
  */
 export async function getDBConnectionAsync(): Promise<Connection> {
-    return createConnection(config);
+    return connection;
```
Why this change?

xianny replied:
`createConnection` actually creates a connection pool. It's intended to be called once per app and continue running in the background. Previous code was creating connection pools every time this method is called.

https://github.com/typeorm/typeorm/blob/master/docs/connection.md

fragosti replied:
Yea but the method was only being called once.  I just don't want to encourage importing this and calling it directly, but rather calling it once and dependency injecting.

xianny replied:
It's kind of confusing to export a function that calls another function that's intended to be only called once. Exporting it encourages it to be called anywhere else.

This way `createConnection` **can** only be called once, and exporting the function as a getter encourages it to be used anywhere else in the app that a connection is needed.

fragosti commented on src/app.ts:
```diff
@@ -0,0 +1,70 @@
+/**
+ * returns the app
+ */
+export async function getAppAsync(config: any = defaultConfig): Promise<express.Application> {
```
might be worth thinking about how to apply this pattern to the other "apps" in the `runners/` directory.

xianny commented on src/services/staking_data_service.ts:
```diff
@@ -120,7 +120,9 @@ export class StakingDataService {
         const currentEpochPoolStats = stakingUtils.getEpochPoolStatsFromRaw(rawCurrentEpochPoolStats[0]);
         const nextEpochPoolStats = stakingUtils.getEpochPoolStatsFromRaw(rawNextEpochPoolStats[0]);
-        const pool7dProtocolFeesGenerated = stakingUtils.getPoolProtocolFeesGeneratedFromRaw(rawPoolSevenDayProtocolFeesGenerated[0]);
+        const pool7dProtocolFeesGenerated = stakingUtils.getPoolProtocolFeesGeneratedFromRaw(
```
just a prettier change

xianny commented on src/index.ts:
```diff
@@ -1,66 +1,17 @@
-        await orderWatcherService.syncOrderbookAsync();
-    } else {
-        logger.warn('API starting without a connection to mesh');
-    }
```
All of this was duplicated in the runners. Moved there and called as a function instead.

xianny commented on src/services/sra_http_service.ts:
```diff
@@ -1,20 +0,0 @@
-        app.use(SRA_PATH, createSRARouter(orderBook));
-        app.use(errorHandler);
-    }
-}
```
changed it to a function instead of a class, and moved it to `/routers`. It's not really a service, it's just setting up routes.

xianny commented on src/services/staking_http_service.ts:
```diff
@@ -1,19 +0,0 @@
-        app.use(STAKING_PATH, createStakingRouter(stakingDataService));
-        app.use(errorHandler);
-    }
-}
```
moved to `/routers`, similar to `sra_http_service.ts`

xianny commented on src/services/swap_http_service.ts:
```diff
@@ -1,20 +0,0 @@
-        app.use(SWAP_PATH, createSwapRouter(swapService));
-        app.use(errorHandler);
-    }
-}
```
moved to `/routers`, same as `sra_http_service.ts`

xianny commented:
Revisited this and realised we needed more refactoring to allow dependency injection of testing mocks.

Changes:
- Cleaned up `/runners` and separated the logic into functions and executions. Removed the duplicate logic in `/index`
- Re-used `/runners` to set up the app in `/app.ts`
- `/index.ts` now uses `/app.ts` to start the API instead of duplicating the logic in `/runners, so we can be sure that all entrypoints are running the same things
- tests use `/app.ts` to inject dependencies and start the API
- removed `/services/*_http_service.ts` because they aren't real services, they just set up routing. Moved the logic to `/routers`.


Questions:
- should we explicitly declare `/index` as the default entry point in package.json? we don't have one right now

fragosti commented on src/runners/http_service_runner.ts:
```diff
@@ -28,41 +32,62 @@ process.on('unhandledRejection', err => {
-    const server = app.listen(config.HTTP_PORT, () => {
-        logger.info(`API (HTTP) listening on port ${config.HTTP_PORT}!\nConfig: ${JSON.stringify(config, null, 2)}`);
+    const server = app.listen(_config.HTTP_PORT, () => {
+        logger.info(`API (HTTP) listening on port ${_config.HTTP_PORT}!\nConfig: ${JSON.stringify(_config, null, 2)}`);
```
Why the underscores?

fragosti commented on src/runners/http_service_runner.ts:
```diff
@@ -28,41 +32,62 @@ process.on('unhandledRejection', err => {
-    let meshClient;
+
+    // staking http service
+    const connection = dependencies.connection || (await getDBConnectionAsync());
```
Shouldn't this be required then?

fragosti commented on src/runners/http_service_runner.ts:
```diff
@@ -28,41 +32,62 @@ process.on('unhandledRejection', err => {
+    // SRA http service
+    const orderBookService = dependencies.orderBookService || new OrderBookService(connection, dependencies.meshClient);
+    configureSRAHttpRouter(app, orderBookService);
+
```
Hmm these too. Shouldn't they be required?

dekz commented on src/runners/http_service_runner.ts:
```diff
@@ -28,41 +32,62 @@ process.on('unhandledRejection', err => {
-    let meshClient;
+
+    // staking http service
+    const connection = dependencies.connection || (await getDBConnectionAsync());
```
I agree, I'd rather it fail if not present than create if undefined.

That way there is consistently one place where all dependencies are created/initialized.

xianny replied:
Removed all of the fallback code. Now there is a `getDefaultAppDependencies` function which is used both for the runners and for index.ts. Defaults depend on the config so we can treat the config as source of truth.

xianny commented on src/routers/sra_router.ts:
```diff
@@ -1,10 +1,23 @@
 
-export const createSRARouter = (orderBook: OrderBookService): express.Router => {
+export const configureSRAHttpRouter = (app: core.Express, orderBook: OrderBookService): void => {
+    app.use(cors());
```
Yeah, this was happening before the refactor too. Was just copying code snippets around. On second thought, all of this seems unnecessary. I'm going to remove all the `configure*` methods.

fragosti replied:
Not 100% what you mean but this architecture is there partially so that you can easily delete code in your own fork of the API (if you only want SRA for example).

dekz commented on src/runners/http_service_runner.ts:
```diff
@@ -28,41 +28,56 @@ process.on('unhandledRejection', err => {
-    const app = express();
+export async function runHttpServiceAsync(
+    dependencies: AppDependencies,
+    config: { HTTP_PORT: string; ETHEREUM_RPC_URL: string },
```
is `ETHEREUM_RPC_URL` needed here if the provider is in the AppDependencies?

xianny replied:
ah good catch. removing!

dekz commented on src/runners/http_service_runner.ts:
```diff
@@ -28,41 +28,56 @@ process.on('unhandledRejection', err => {
+    app.use(errorHandler);
+
+    // websocket service
+    if (dependencies.meshClient) {
```
Since we require `meshClient` for SRA `POST /orders` this should always be present. Should we fail if this doesn't exist? Or is this helpful for testing?

xianny replied:
Yup this is helpful for testing. Previous implementation also had this if-clause and I preserved the behaviour because I didn't have a good reason to change it... **should** we fail if this doesn't exist? @fragosti thoughts?

fragosti replied:
Depends what you mean by fail. We need to balance keeping as many systems working as possible while also making it very visible when something fails.

I would keep the try-catches but make sure we `logger.error` and then alert on that?

dekz commented on src/runners/order_watcher_service_runner.ts:
```diff
@@ -1,19 +1,38 @@
+if (require.main === module) {
+    (async () => {
+        const provider = providerUtils.createWeb3Provider(defaultConfig.ETHEREUM_RPC_URL);
+        const { connection, meshClient } = await getDefaultAppDependenciesAsync(provider, defaultConfig);
```
thoughts on creating the provider in `getDefaultAppDependenciesAsync` ?

xianny replied:
I had that at first, but ran into trouble with the provider for Ganache :/

fragosti commented on src/runners/http_service_runner.ts:
```diff
@@ -28,41 +28,56 @@ process.on('unhandledRejection', err => {
+    if (dependencies.swapService) {
+        app.use(SWAP_PATH, createSwapRouter(dependencies.swapService));
+    } else {
+        logger.warn(`API running without swap service`);
```
`logger.error` ?

xianny replied:
It was warn before but I can change it to error :)

xianny commented on src/handlers/root_handler.ts:
```diff
@@ -1,4 +1,3 @@
-
```
prettier
